File: backend/src/TaskMgmt.Domain/Entities/TaskHistory.cs
using TaskMgmt.Domain.Common;
using TaskMgmt.Domain.Enums;

namespace TaskMgmt.Domain.Entities;

// Bảng chỉ-ghi (append-only): không kế thừa AuditableEntity vì lịch sử không bao giờ được sửa
// sau khi tạo - không có UpdatedAtUtc/UpdatedByUserId vì khái niệm "sửa lịch sử" không tồn tại.
public class TaskHistory : BaseEntity
{
    public required Guid WorkTaskId { get; set; }
    public WorkTask? WorkTask { get; set; }

    public TaskHistoryActionType ActionType { get; set; }
    public required string Description { get; set; }

    public string? FieldName { get; set; }
    public string? OldValue { get; set; }
    public string? NewValue { get; set; }

    // Dùng cho sự kiện liên quan tới người tham gia (assignee added/removed/role changed).
    public Guid? TargetUserId { get; set; }
    public User? Target { get; set; }

    public Guid? ActorUserId { get; set; }
    public User? Actor { get; set; }

    public DateTimeOffset CreatedAtUtc { get; set; }
}

File: mobile/taskmgmt_app/lib/features/task_histories/domain/entities/task_history.dart
enum TaskHistoryActionType {
  created,
  fieldChanged,
  statusChanged,
  deleted,
  assigneeAdded,
  assigneeRemoved,
  assigneeRoleChanged,
  commentAdded,
  attachmentAdded,
  attachmentRemoved,
}

TaskHistoryActionType taskHistoryActionTypeFromString(String value) => switch (value) {
      'Created' => TaskHistoryActionType.created,
      'FieldChanged' => TaskHistoryActionType.fieldChanged,
      'StatusChanged' => TaskHistoryActionType.statusChanged,
      'Deleted' => TaskHistoryActionType.deleted,
      'AssigneeAdded' => TaskHistoryActionType.assigneeAdded,
      'AssigneeRemoved' => TaskHistoryActionType.assigneeRemoved,
      'AssigneeRoleChanged' => TaskHistoryActionType.assigneeRoleChanged,
      'CommentAdded' => TaskHistoryActionType.commentAdded,
      'AttachmentAdded' => TaskHistoryActionType.attachmentAdded,
      'AttachmentRemoved' => TaskHistoryActionType.attachmentRemoved,
      _ => TaskHistoryActionType.fieldChanged,
    };

class TaskHistory {
  const TaskHistory({
    required this.id,
    required this.workTaskId,
    required this.actionType,
    required this.description,
    required this.fieldName,
    required this.oldValue,
    required this.newValue,
    required this.actorUserId,
    required this.actorFullName,
    required this.actorEmail,
    required this.targetUserId,
    required this.targetUserFullName,
    required this.targetUserEmail,
    required this.createdAtUtc,
  });

  final String id;
  final String workTaskId;
  final TaskHistoryActionType actionType;
  final String description;
  final String? fieldName;
  final String? oldValue;
  final String? newValue;
  final String? actorUserId;
  final String actorFullName;
  final String actorEmail;
  final String? targetUserId;
  final String? targetUserFullName;
  final String? targetUserEmail;
  final DateTime createdAtUtc;
}
